bundles/jcr/resource/src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrItemResourceFactory.java

```diff
@@ -78,17 +78,15 @@ public JcrItemResourceFactory(Session session, HelperData helper) {
         }
 
         Node parentNode = null;
+        String parentResourcePath = null;
         if (parent != null) {
             parentNode = parent.adaptTo(Node.class);
+            parentResourcePath = parent.getPath();
         }
 
         Item item = null;
-        if (parentNode != null && jcrPath.startsWith(parentNode.getPath())) {
-            final String parentJcrPath = parentNode.getPath();
-            String subPath = jcrPath.substring(parentJcrPath.length());
-            if (!subPath.isEmpty() && subPath.charAt(0) == '/') {
-                subPath = subPath.substring(1);
-            }
+        if (parentNode != null && resourcePath.startsWith(parentResourcePath)) {
+            String subPath = resourcePath.substring(parentResourcePath.length() + 1);
```

Comment:
@trekawek I have partially reverted the change so that it still avoids the getPath call but should also work with direct children of the root.

At the moment, it is very hard to implement tests for these optimisations since the provider has access to the resolver through the ResolveContext and therefore one has to setup a resolver with the provider. This, however, is very difficult because the ResourceResolverImpl requires a CommonResourceResolverFactoryImpl which requires a ResourceResolverFactoryActivator which depends on OSGi.

In my opinion, there should be a way to instantiate a resolver for testing purposes from a list of providers.